sample.go

```diff
@@ -0,0 +1,215 @@
+	"unicode"
+
+	"gopkg.in/yaml.v2"
+)
+
+// SampleYAML writes YAML output to w, indented by indent spaces
+// that holds the attributes in attrs with descriptions found
+// in the given fields. An entry for any attribute in fields not
+// in attrs will be generated but commented out.
+func SampleYAML(w io.Writer, indent int, attrs map[string]interface{}, fields Fields) error {
+	indentStr := strings.Repeat(" ", indent)
+	orderedFields := make(fieldsByGroup, 0, len(fields))
+	for name, f := range fields {
+		orderedFields = append(orderedFields, attrWithName{
+			name: name,
+			Attr: f,
+		})
+	}
+	sort.Sort(orderedFields)
+	for i, f := range orderedFields {
+		if i > 0 {
+			fmt.Fprintf(w, "\n")
+		}
+		writeSampleDescription(w, f.Attr, indentStr+"# ")
+		val, ok := attrs[f.name]
+		if ok {
+			fmt.Fprintf(w, "%s:", f.name)
+			indentVal(w, val, indentStr)
+		} else {
+			if f.Example != nil {
+				val = f.Example
+			} else {
+				val = sampleValue(f.Type)
+			}
+			fmt.Fprintf(w, "# %s:", f.name)
+			indentVal(w, val, indentStr+"# ")
+		}
+	}
+	return nil
+}
+
+const textWidth = 80
+
+func writeSampleDescription(w io.Writer, f Attr, indent string) {
+	previousText := false
+
+	// section marks the start of a new section of the comment;
+	// sections are separated with empty lines.
+	section := func() {
+		if previousText {
+			fmt.Fprintf(w, "%s\n", strings.TrimRightFunc(indent, unicode.IsSpace))
+		}
+		previousText = true
+	}
+
+	descr := strings.TrimSpace(f.Description)
+	if descr != "" {
+		section()
+		if !strings.HasSuffix(descr, ".") {
+			descr += "."
+		}
+		doc.ToText(w, descr, indent, "    ", textWidth-len(indent))
+	}
+	vars := make([]string, 0, len(f.EnvVars)+1)
+	if f.EnvVar != "" {
+		vars = append(vars, "$"+f.EnvVar)
+	}
+	for _, v := range f.EnvVars {
+		vars = append(vars, "$"+v)
+	}
+	if len(vars) > 0 {
+		section()
+		fmt.Fprintf(w, "%sDefault value taken from %s.\n", indent, wordyList(vars))
+	}
+	attrText := ""
+	switch {
+	case f.Secret && f.Immutable:
+		attrText = "immutable and considered secret"
+	case f.Secret:
+		attrText = "considered secret"
+	case f.Immutable:
+		attrText = "immutable"
+	}
+	if attrText != "" {
+		section()
+		fmt.Fprintf(w, "%sThis attribute is %s.\n", indent, attrText)
+	}
+	section()
+}
+
+func emptyLine(w io.Writer, indent string) {
+	fmt.Fprintf(w, "%s\n", strings.TrimRightFunc(indent, unicode.IsSpace))
+}
+
+func wordyList(words []string) string {
+	if len(words) == 0 {
+		return ""
+	}
+	if len(words) == 1 {
+		return words[0]
+	}
+	return strings.Join(words[0:len(words)-1], ", ") + " or " + words[len(words)-1]
+}
+
+var groupPriority = map[Group]int{
+	ProviderGroup: 3,
+	AccountGroup:  2,
+	EnvironGroup:  1,
+}
+
+type attrWithName struct {
+	name string
+	Attr
+}
+
+type fieldsByGroup []attrWithName
+
+func (f fieldsByGroup) Len() int {
+	return len(f)
+}
+
+func (f fieldsByGroup) Swap(i, j int) {
+	f[i], f[j] = f[j], f[i]
+}
+
+func (f fieldsByGroup) Less(i0, i1 int) bool {
```

Comment:
I think I'd like to see the same parameters for Less as Swap. I don't mind which ones you choose.